pkg/pillar/pubsub/socketdriver/publish.go

```diff
@@ -116,9 +133,49 @@ func (s *Publisher) Load() (map[string][]byte, int, error) {
 		}
 		items[key] = sb
 	}
+
+	// Try to recover lost items (have backup but missing the original).
+	if s.persistent {
+		const backupSuffix = ".json.bak"
+		var keysToRecover []string
+		for _, file := range files {
+			if strings.HasSuffix(file.Name(), backupSuffix) {
+				key := strings.TrimSuffix(file.Name(), backupSuffix)
+				if _, loaded := items[key]; !loaded {
+					keysToRecover = append(keysToRecover, key)
+				}
+			}
+		}
+		for _, key := range keysToRecover {
+			sb, err := s.RecoverFromBackup(key)
+			if err != nil {
+				s.log.Errorf("Failed to recover %s/%s: %v", s.name, key, err)
+				continue
+			}
+			items[key] = sb
+			s.log.Warnf("Using backup of %s/%s", s.name, key)
+		}
+	}
 	return items, restartCounter, err
 }
 
+// RecoverFromBackup tries to recover possibly corrupted item from a backup file.
+// Typically called when loaded value cannot be unmarshalled into a structure,
```

Comment:
Shouldn't we try to unmarshal somewhere above or determine that the file is just empty?

Reply:
Good point @giggsoff 
I don't know if the caller to populate does a json.Unmarshal or if that waits until someone does e.g., a Get().

For this we'd want to catch empty or truncated files by doing an unmarshall either here or in the caller if it already does an unmarshal.

Reply:
Note that the caller does a ParseTemplate first or the original and then of the backup so we're good on the unmarshal check.